policy: Policy Map Overflow Lockdown

nathanjsweet opened the pull request:
see commits for details.
```release-note
endpoint: Add an option to lock endpoints down (that is, drop all traffic) when their policy maps overflow.
```

nathanjsweet committed:
policy: Add SupportedProtocols method API

Add method to return all transport protocols supported
by the policy engine for other packages to consume.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
helm: Add Helm Option for Policy Map Overflow

"endpointLockdownOnMapOverflow" is the option
that maps to config option "enable-policy-overflow-endpoint-lockdown".

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
option: Add Option to Manage Policy Map Overflow Behavior

Add an option (`enable-lockdown-endpoint-on-policy-overflow`),
set to "false" by default.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
policy: Policy Map Overflow Lockdown

Assuming `enable-lockdown-endpoint-on-policy-overflow`)
is set to "true" (by default it is "false"),when a
policy map needs more entries than is allocated for its
bpf map, the endpoint is locked down. The policy cache is
reset and the policy map is populated with a set of default
deny entries. The endpoint is put into full default-deny
for every protocol (except for ICMP).

If the lockdown calls fail, then the endpoint will be
regenerated, and another lockdown will be attempted (until
the underlying map pressure is relieved).

Incremental deletion is now done first in cases where
incremental addition before deletion would overflow the
policy map.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

nathanjsweet committed:
docs: Update Endpoint Documentation

Update Endpoint documentation to include lockdown
description.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1306,14 +1398,61 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+	// 5. Update the pressure metric, because other code paths
+	//    to updating it are now disabled in lockdown mode.
+	// 6. Update the state of the endpoint
+	err = e.policyMap.Unpin()
```
nit: if err := e.policyMap.Unpin(); err != nil {...}

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,6 +1302,59 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+			logfields.EndpointID: e.ID,
+		}).Warn("Policy Map exceeds the policy map max entries limit, locking endpoint down...")
+
+		err := e.endpointPolicyLockdown()
```
nit: if err := e.endpointPolicyLockdown(); err != nil {...}

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1306,14 +1398,61 @@ func (e *Endpoint) applyPolicyMapChanges() error {
 	// Apply pending policy map changes first so that desired map is up-to-date before
 	// we diff the maps below.
-	err := e.applyPolicyMapChanges()
+	err := e.applyPolicyMapChangesLocked()
```
nit: if err := xxx; err != nil {...}

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1434,8 +1573,17 @@ func (e *Endpoint) syncPolicyMapWithDump() error {
 	// Apply pending policy map changes first so that desired map is up-to-date before
 	// we diff the maps below.
-	err := e.applyPolicyMapChanges()
+	err := e.applyPolicyMapChangesLocked()
```
nit: same as above^

derailed commented on pkg/endpoint/bpf.go:
```diff
@@ -1249,6 +1302,59 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.EndpointID: e.ID,
+		}).Warn("Policy Map exceeds the policy map max entries limit, locking endpoint down...")
```
nit: Perhaps we could also notify users that this is happening since EnableEndpointLockdownOnPolicyOverflow is set?

derailed reviewed (approved):
@nathanjsweet Nice work!

learnitall commented on pkg/endpoint/bpf.go:
```diff
@@ -1156,14 +1208,15 @@ func (e *Endpoint) ApplyPolicyMapChanges(proxyWaitGroup *completion.WaitGroup) e
+// applyPolicyMapChangesLocked applies any incremental policy map changes
+// collected on the desired policy. The method must be called with the
+// endpoint mutex held. It returns two special errors that must be considered,
+// "ErrPolicyEntryMaxExceeded".
```
Nit: the comment says: "it returns two special errors that must be considered" but only one is listed.

learnitall commented on pkg/endpoint/bpf.go:
```diff
@@ -1174,6 +1227,59 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		e.getLogger().WithFields(logrus.Fields{
+			logfields.EndpointID: e.ID,
+		}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+			option.EnableEndpointLockdownOnPolicyOverflow)
```
Would it be possible to add metrics to add a method for tracking when an endpoint is locked that doesn't require watching logs?

nathanjsweet replied:
We shouldn't do this. Policy map pressure can already provide the same information. I'll grant that it seems like "something" has occurred, but that is just that map pressure has hit "1.0". A metric behind a config flag can cause problems as well and is another point against creating a metric for this. I will make sure to add documentation to the section on the flag to encourage good map pressure monitoring

learnitall replied:
Sounds good, thanks!

learnitall reviewed (approved):
Non-blocking feedback, just a nit and a question about a potential next step. LGTM! Thanks for all the comments.

nathanjsweet committed:
endpoint: Update Policy Map Pressure on Lockdown

If an endpoint is in lockdown it should publish an
accurate policy map pressure above "1.0". This makes
the value actionable in prometheus.

Update lockdown documentation to explicitly reference
the alertability of the metric.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

rgo3 commented on pkg/endpoint/bpf.go:
```diff
@@ -1174,6 +1227,59 @@ func (e *Endpoint) applyPolicyMapChanges() error {
+		}).Warnf("The policy map exceeds the max entries limit, %s is enabled, locking the endpoint down.",
+			option.EnableEndpointLockdownOnPolicyOverflow)
+
+		if err := e.endpointPolicyLockdown(); err != nil {
```
Does a call to this function need to happen if the endpoint is already locked? If not we should wrap it in an `if !e.lockdown {}` block or `endpointPolicyLockdown()` could return early `if e.lockdown == true`. Otherwise, IIUC, we keep recreating the map to insert the deny all entries.

nathanjsweet replied:
> Does a call to this function need to happen if the endpoint is already locked?

Yes it does.

> Otherwise, IIUC, we keep recreating the map to insert the deny all entries.

Correct.

rgo3 replied:
Could you elaborate why is it necessary to do this if the endpoint is already locked down? Is it to accommodate agent restart and upgrade scenarios?

nathanjsweet replied:
No that's a good point. Fixed.

rgo3 reviewed (changes requested):
Otherwise LGTM!

ldelossa reviewed (approved):
I am not that versed on the Policy specifics, but from an datapath/eBPF interface point of view, the changes look good to me.

squeed commented on pkg/maps/policymap/policymap.go:
```diff
@@ -471,6 +471,13 @@ func Create(path string) error {
 }
 
+// Create unpinned creates a policy map without pinning it to the filesystem
+func CreateUnpinned(path string) (*PolicyMap, error) {
```
Why would we want to do this? What does the path mean if not pinned?

rgo3 replied:
I think technically what is called `path` here is/will be the name of the map, but almost all functions in this file/pkg use this wording, so it would probably be more confusing to change it in this instance.

nathanjsweet replied:
Unpinned maps still ned a path name for [the register](https://github.com/cilium/cilium/blob/main/pkg/bpf/map_register_linux.go#L20).

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,62 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+	var deleteLen int
+	for k := range changes.Deletes {
+		if _, ok := changes.Adds[k]; !ok {
+			deleteLen++
```
Suggestion: add `ChangeState.Size()` (or something) that computes the actual change in MapState entries.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,62 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+			e.getLogger().WithFields(logrus.Fields{
+				logfields.EndpointID: e.ID,
+			}).WithError(err).Error("Failed to come out of lockdown successfully:" +
+				"The endpoint may be compromised, consider quarantining or shutting down this node.")
```
nit: compromised is *not* a good word to use here :-)

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,62 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+	// insertion if we proceed).
+	if e.desiredPolicy.GetPolicyMap().Len()+len(changes.Adds)-deleteLen > policymap.MaxEntries {
+		// Revert the map changes that were just consumed.
+		e.desiredPolicy.RevertChanges(changes)
```
Are we concerned with lots and lots of changes building up? Would it be better to apply the changes to the userspace policy map and do a full sync when coming out of lockdown?

nathanjsweet replied:
Further down in the method other logic relies on `ChangeState`. For example, checking if there were changes to update envoy redirects. Lockdown bypasses all of that logic. Once the endpoint is remediated the logic after the `lockdown` logic would be oblivious to any of the changes.

squeed replied:
But those don't matter, right? If we switch to the lockdown map, no traffic is going to Envoy...

nathanjsweet replied:
They don't matter during lockdown, but after coming out of lockdown that envoy logic will be skipped.

squeed replied:
I'm still not sure this is what we want; RevertChanges doesn't re-queue the changes, it just backs them out. So the incremental changes are lost regardless.

I'm afraid this could cause us to flap in and out of lockdown, where an incremental update is rejected, keeping desiredPolicy.MapState under the threshold.

nathanjsweet replied:
Alright. I removed it. I was worried about the envoy logic not being triggered, but I realized I was wrong. Of course, it will be, because an empty changeset can't bring the endpoint out of lockdown.

rgo3 commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,63 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+		return ErrPolicyEntryMaxExceeded
+	}
+
+	if e.lockdown {
```
Rather late observation: what happens to the pinned lockdown map when we are coming out of lockdown? How does that get cleaned up/ replaced by a new "regular" policy map?

nathanjsweet replied:
The `endpoint.policyMap` is replaced in lockdown, given a temporary pin path, and then pinned to the correct path for the datapath. Coming out of lockdown `syncPolicyMap` syncs `desiredPolicy` (which has all of the policies) and `realizedPolicy` (which, during lockdown, has 6 deny all policies in it). The logic inserts all of the `desiredPolicy` entries into `realizedPolicy` and the change state is inserted into the datapath map. No clean up is necessary. The reason for replacing the "old" map for lockdown is that it is faster to simply replace the map rather than do delete syscalls for all of the entries (especially for large maps). The inverse is not true; when coming out of lockdown the map must be rebuilt, so no replacement swap can work to make it faster, the entries must be reinserted. I would like to start utilizing some of the batch ops in this package (which would definitely help in this situation).

rgo3 replied:
Ok I think I follow: if we come out of lockdown we reuse the lockdown map and fill it with actual policies correct? But then the name of the map will be wrong right? It will have the `_lockdown` suffix since we are not recreating the policy map, and the lockdown mapped was named with the tmp pin path.

nathanjsweet replied:
No. During lockdown it is created `Unpinned` with a `_lockdown` suffix, but when it is pinned the file is changed (on line 1199) `e.policyMap.Pin(e.policyMapPath())` so it has the correct name during lockdown (otherwise it wouldn't work).

rgo3 replied:
Sorry for the long wait... I think I am talking about the name stored with the bpf ressource, and you are talking about the pin path. I understand the pin path handling is correct, however the map name will keep the `_lockdown` suffix if its not replaced. My suggestion would be that we just don't use the `_lockdown` suffix at all as it doesn't really have any use IIUC. Unfortunately, I also can't prove my point using evidence, `bpftool `never displays more than 15 characters of a map name. I will approve regardless to not block this PR any longer and I don't think it has any bad side effects, but I am pretty sure we are permanently changing the map name when locking down an endpoint, even if the pin path stays the same.

nathanjsweet committed:
policy: Add Size Method to ChangeState

Add a Size method to ChangeState so that the
overall impact of a ChangeState can be utilized
by the endpoint package.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,58 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+				"The endpoint may be compromised, consider quarantining or shutting down this node.")
+			return fmt.Errorf("Failed to come out of lockdown successfully: %w", err)
+		}
+		e.realizedPolicy = e.desiredPolicy
```
There is other code that compares `e.realizedPolicy` and `e.desiredPolicy`; will this have an effect?

nathanjsweet replied:
`realizedPolicy` and `desiredPolicy` are compared to determine if `syncPolicyMap...` has been called. Their difference indicates regeneration or syncing is taking place. They are equal when their contents are the same (which they are in this case).

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -962,6 +1014,58 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+			e.lockdown = true
+		}
+		e.updatePolicyMapPressureMetric(float64(e.desiredPolicy.GetPolicyMap().Len() + changeSize))
+		return ErrPolicyEntryMaxExceeded
```
This will cause regeneration of this endpoint to permanently fail. That is probably not desired; there may be code that waits for all endpoints to be regenerated.

Rather, we will need to find a way to cleanly transition an endpoint in to lockdown mode that does not otherwise block regeneration.

nathanjsweet replied:
> This will cause regeneration of this endpoint to permanently fail. That is probably not desired; there may be code that waits for all endpoints to be regenerated.

Good catch.

>Rather, we will need to find a way to cleanly transition an endpoint in to lockdown mode that does not otherwise block regeneration.

I updated `runPreCompilationSteps` to swallow the error. No code in `runPreCompilationSteps` will be affected thereafter. I also updated `regenerateBPF` to avoid syncing `desiredPolicy` into `realizedPolicy` if the endpoint is in lockdown. Both those changes together will allow regeneration to proceed without affecting the lockdown state.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -636,8 +657,8 @@ func (e *Endpoint) runPreCompilationSteps(regenContext *regenerationContext) (pr
-		err = e.applyPolicyMapChanges(regenContext, e.desiredPolicy != e.realizedPolicy)
-		if err != nil {
+		err = e.applyPolicyMapChangesLocked(regenContext, e.desiredPolicy != e.realizedPolicy)
+		if err != nil && errors.Is(err, ErrPolicyEntryMaxExceeded) {
```
:thinking: so this means we proceed on other errors, e.g. bpf writes?

nathanjsweet replied:
Oops. The opposite. We need to swallow the error. I made this mistake by fixing a linting error incorrectly.

pippolo84 commented on pkg/bpf/map_linux.go:
```diff
@@ -469,17 +475,17 @@ func (m *Map) OpenOrCreate() error {
 //
 // TODO(tb): Remove this when all map creation takes MapSpec.
-func (m *Map) CreateUnpinned() error {
+func (m *Map) CreateUnpinned(register bool) error {
```
Nit: should we add a brief explanation of the `register` parameter in the method docstring?

pippolo84 commented on pkg/endpoint/bpf.go:
```diff
@@ -1051,32 +1118,22 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
-	// Add policy map entries before deleting to avoid transient drops
+	// Add policy map entries before deleting to avoid transient drops. If there
+	// isn't enough space to add all the entries before deleting some, then delete
+	// first.
```
Question: should we log something (maybe at a warning level) when we invert the insertion/deletion order? Or at least clarify in the docs that when we are close to the limit (but not yet in the condition to lock the endpoint) some transient drops might happen due to this inversion?

pippolo84 reviewed (approved):
LGTM! :100: 

Left a non blocking nit and a question inline.

squeed commented on pkg/endpoint/bpf.go:
```diff
@@ -1108,13 +1168,127 @@ func (e *Endpoint) applyPolicyMapChanges(regenContext *regenerationContext, hasN
+// false if it did not.
+func (e *Endpoint) stopLockdownLocked() bool {
+	if e.lockdown {
+		e.forcePolicyCompute = true
```
Good plan :-)
